src/main/java/io/vertx/core/spi/json/JsonCodec.java

```diff
@@ -14,12 +14,25 @@
 import io.vertx.core.buffer.Buffer;
 import io.vertx.core.json.DecodeException;
 import io.vertx.core.json.EncodeException;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
 
 /**
  * @author <a href="mailto:julien@julienviet.com">Julien Viet</a>
  */
 public interface JsonCodec {
 
+  /**
+   * Decode a given JSON string.
+   *
+   * @param json the JSON string.
+   * @return a JSON element which can be a {@link JsonArray}, {@link JsonObject}, {@link String}, ...etc. if the string contains an array, object, string, ...etc
+   * @throws DecodeException when there is a parsing or invalid mapping.
+   */
+  default Object fromString(String json) throws DecodeException {
```

Comment:
is it a required change to the PR or is it only an accommodation change ? I am not sure of this

Reply:
It is a required change. This new method complies to the contract of `io.vertx.core.json.Json#decodeValue(java.lang.String)`, which is returning a `JsonArray`, `JsonObject`, `String, ...etc. if the content is an array, object, string, ...etc. in all cases.

The existing methods of `DatabindCodec` do not comply to the contract if, for example, the `ObjectMapper` is configured with the Jackson Scala module.

Reply:
why isn't `io.vertx.core.json.Json#decodeValue(java.lang.String)` calling `fromString(Object.class)` incorrect ?

Reply:
Because, with the DatabindCodec, if the object mapper has been configured with the Jackson Scala Module, the type resolved for a json string or json array is not a Java Map or List, but a Scala Map or Sequence.

Reply:
it seems that it only then works because of this 

```java
  @Override
  public Object fromString(String str) throws DecodeException {
    return JacksonCodec.fromParser(createParser(str), Object.class);
  }
```

which makes this `fromString` behavior very much different than the other `fromString` overloads in the `DatabindCodec` class

Reply:
Right, the behavior it's different, it's the only one which has the contract of returning a valid Vert.x JSON object: `String`, `Boolean`, `JsonObject`, `JsonArray`.

Users are not supposed to work directly with codec objects, they are supposed to work with the `Json` class. This one has two different methods (and more): `decodeValue(String)` and `decodeValue(String, Class<T>)`. The first one has the contract of returning a valid Vert.x JSON object, the other just says it will try to deserialize something with the type of parameter. It works as we expect with the `JacksonCodec` but, with the `DatabindCodec`, it may not work if the object mapper is configured in a way we hadn't anticipated.